crc-fast-gen/src/lib.rs

```diff
@@ -310,21 +320,175 @@ unsafe fn fold_by_4_128(
+
+    // Left shift by 4 bytes using tbl lookup. Out-of index automatically zeroes the right bytes.
+    let left_shift_4bytes_mask = vld1q_u8([255, 255, 255, 255, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].as_ptr());
+    let x = vandq_u8(
+        veorq_u8(
+            vqtbl1q_u8(x, left_shift_4bytes_mask),
+            k5mul,
+        ),
+        vreinterpretq_u8_s32(vld1q_s32([!0, !0, !0, 0].as_ptr())),
+    );
+
+    let k6mul = core::mem::transmute(vmull_p64(vgetq_lane_p64(vreinterpretq_p64_u8(x), 1), vgetq_lane_p64(vreinterpretq_p64_u8(k5k6), 1)));
+    let x = vandq_u8(veorq_u8(x, k6mul), vreinterpretq_u8_s32(vld1q_s32([!0, !0, 0, 0].as_ptr())));
+
+    let right_shift_4bytes_mask = vld1q_u8([4,5,6,7,8,9,10,11,12,13,14,15,255,255,255,255].as_ptr());
+    let pu = vreinterpretq_u8_s64(vld1q_s64([Q_X, U].as_ptr()));
+    let t1 = core::mem::transmute(
+        vmull_p64(
+            vgetq_lane_p64(vreinterpretq_p64_u8(vqtbl1q_u8(x, right_shift_4bytes_mask)), 0),
+            vgetq_lane_p64(vreinterpretq_p64_u8(pu), 1),
+        ),
+    );
+    let t2 = core::mem::transmute(
+        vmull_p64(
+            vgetq_lane_p64(vreinterpretq_p64_u8(vqtbl1q_u8(t1, right_shift_4bytes_mask)), 0),
+            vgetq_lane_p64(vreinterpretq_p64_u8(pu), 0),
+        ),
+    );
+
+    let x = veorq_u8(x, t2);
+    let mut c = vgetq_lane_u32(vreinterpretq_u32_u8(x), 0) as u32;
+
+"# +
+     format!("    c = c >> {};", 32 - n).as_str() +
+     format!("    c ^ {}", output_xor).as_str() +
+r#"
+}
+
+#[cfg(target_arch = "aarch64")]
+#[target_feature(enable = "aes")]
+unsafe fn fold_by_4_128_aarch64(
+    x3: uint8x16_t,
+    x2: uint8x16_t,
+    x1: uint8x16_t,
+    x0: uint8x16_t,
+    k1k2: uint8x16_t,
+    shuf_mask: uint8x16_t,
+    octets: &mut &[u8],
+) -> (uint8x16_t, uint8x16_t, uint8x16_t, uint8x16_t) {
+    let y3 = vld1q_u8(octets.as_ptr());
+    *octets = &octets[16..];
+    let y2 = vld1q_u8(octets.as_ptr());
+    *octets = &octets[16..];
+    let y1 = vld1q_u8(octets.as_ptr());
+    *octets = &octets[16..];
+    let y0 = vld1q_u8(octets.as_ptr());
+    *octets = &octets[16..];
+
+    let y3 = vqtbl1q_u8(y3, shuf_mask);
+    let y2 = vqtbl1q_u8(y2, shuf_mask);
+    let y1 = vqtbl1q_u8(y1, shuf_mask);
+    let y0 = vqtbl1q_u8(y0, shuf_mask);
+
+    let x3 = reduce128_aarch64(x3, y3, k1k2);
+    let x2 = reduce128_aarch64(x2, y2, k1k2);
+    let x1 = reduce128_aarch64(x1, y1, k1k2);
+    let x0 = reduce128_aarch64(x0, y0, k1k2);
+    (x3, x2, x1, x0)
+}
+
+#[cfg(target_arch = "aarch64")]
+#[target_feature(enable = "aes")]
+unsafe fn reduce128_aarch64(a: uint8x16_t, b: uint8x16_t, keys: uint8x16_t) -> uint8x16_t {
+    let t1 = core::mem::transmute(vmull_p64(vgetq_lane_p64(vreinterpretq_p64_u8(a), 0), vgetq_lane_p64(vreinterpretq_p64_u8(keys), 1)));
+    let t2 = core::mem::transmute(vmull_p64(vgetq_lane_p64(vreinterpretq_p64_u8(a), 1), vgetq_lane_p64(vreinterpretq_p64_u8(keys), 0)));
```

Comment:
TODO: align high-high and lo-lo, verify it can use the `PMULL2` instruction under the hood